document-store/src/main/java/org/hypertrace/core/documentstore/Collection.java

```diff
@@ -80,6 +81,24 @@ public interface Collection {
    */
   CloseableIterator<Document> aggregate(final org.hypertrace.core.documentstore.query.Query query);
 
+  /**
+   * Reads the first document matching the filter and sorting criteria given in the {@param query},
+   * updates the document as specified in {@param updateDocument} and returns the document (if
+   * exists) including the fields selected in the {@param query}
+   *
+   * @return The document before updating if exists, otherwise an empty optional
+   * @throws IOException if there was any error in updating/fetching the document
+   * @implSpec The definition of an update here is
+   *     <ol>
+   *       <li>The existing fields/sub-documents will be updated
+   *       <li>New fields/sub-documents will be created if they do not exist
+   *       <li>None of the existing fields/sub-documents will be removed
+   *     </ol>
+   */
+  Optional<Document> atomicReadAndUpdateDocument(
```

Comment:
if we could align name with other APIs - `updateAndReturnFirst` or `updateAndReturn`?  Or `update` should be fine?

we have that has a similar use case.
```UpdateResult update(Key key, Document document, Filter condition) throws IOException;```

So, this can be :
```update(final org.hypertrace.core.documentstore.query.Query query, final Document updateDocument) throws IOException;```